### PR TITLE
prevent underflow in parse_number on check for Hex, Octal, or Binary

### PR DESCRIPTION
**Pull request by CDanU**
pc_length = pc_temp.len(); ---> 0
ch        = pc_temp.str[pc_length - 1];

changes some length return types to size_t

--------------------------------------------

While figuring out if the [operator[]](https://github.com/uncrustify/uncrustify/blob/7f268aff6c652be6088fd619d5fccb05bb04f509/src/unc_text.h#L164) function in unc_text.h actually should or should not take in negative param values I set up a conditional breakpoint that triggers when [pc_length](https://github.com/uncrustify/uncrustify/blob/e4e0cc7515b26d464dc11f4d42e85aa88f9ce9e9/src/tokenize.cpp#L684) == 0 in parse_number in _tokenize.cpp_ while uncrustifying _uncrustify.cpp_. 

In `ch        = pc_temp.str[pc_length - 1];` 
`pc_length - 1` becomes -1 and underfloats the _size_t_ of [operator[]](https://github.com/uncrustify/uncrustify/blob/7f268aff6c652be6088fd619d5fccb05bb04f509/src/unc_text.h#L164) and the (idx < m_chars.size()) check returns false.

If the input type param of [operator[]](https://github.com/uncrustify/uncrustify/blob/7f268aff6c652be6088fd619d5fccb05bb04f509/src/unc_text.h#L164) is not _size_t_ but _int_ ( as it was before 7f268aff6c652be6088fd619d5fccb05bb04f509) the comparison of int(-1) and ulong(m_chars.size()) also returns false.

Seeing the comment 

> /* either just 0 or 0.1 or 0UL, etc */

 further down below in the parse_number function I get the feeling that the comparison of negative _int_ and _ulong_ was actually intended to return false, i.e. it was not a mere coincidence.

Still I think _size_t_ should be used here as it makes more clear what values are appropriate.

### CODE REVIEW

File: src/tokenize.cpp [hunk 1]
```diff
@@ -682,7 +682,7 @@ static bool parse_number(tok_ctx &ctx, chunk_t &pc)
          pc_temp.str.append(ch);
       }
       pc_length = pc_temp.len();
-      ch        = pc_temp.str[pc_length - 1];
+      ch        = (pc_length > 0) ? pc_temp.str[pc_length - 1] : -1;
```
Comment (mihaipopescu) [inline src/tokenize.cpp]: `pc_length > 0` is always true because of it's type.

File: src/unc_text.h [hunk 2]
```diff
@@ -67,7 +67,7 @@ class unc_text
 
 
    /* grab the number of characters */
-   size_t size() const
+   value_type::size_type size() const
```
Comment (mihaipopescu) [inline src/unc_text.h]: why is this necessary ? I don't really get it.

**Comment by mihaipopescu** [comment]
Thanks for the bug/patch! Can you also add a test case to prove the bug ?

File: src/tokenize.cpp (hunk 1 above)
Comment (CDanU) [inline src/tokenize.cpp]: How so ? Without my commit the type of  pc_length is int, pc_temp.len(); returns a size_t which can be 0. In the while loop CharTable::IsKw2 can fail right from the beginning.

**Comment by gmaurel** [comment]
Check again the *.h files

**Comment by gmaurel** [comment]
The problem is solved by PR https://github.com/uncrustify/uncrustify/pull/827
